src/types/mod.rs

```diff
@@ -0,0 +1,811 @@
+// Rust Simplicity Library
+// Written in 2023 by
+//   Andrew Poelstra <apoelstra@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Types and Type Inference
+//!
+//! Every Simplicity expression has two types associated with it: a source and
+//! a target. These types are inferred from the structure of the program, with
+//! the additional rule that any unbound types in a complete program are simply
+//! set to the unit type.
+//!
+//! Concrete are one of three things
+//!   * A unit type, which has one value
+//!   * A sum of two other types
+//!   * A product of two other types
+//!
+//! During type inference, types may also be *free*, which means that they may
+//! turn out to be any type. Additional constraints are provided by the program
+//! itself, and type inference proceeds by using the union-bound algorithm to
+//! unify all the constraints on each type.
+//!
+//! If unification succeeds, types may then be finalized, which sets all free
+//! types to unit and does one additional check (the "occurs check") which
+//! makes sure there are no infinitely-sized types, which are illegal in
+//! Simplicity.
```

Comment:
This is confusing. I think the narrative should be that "types" may be variable and incomplete. Finalized types are types without free subtypes.

Types may be free, bound to some structure (unit, sum of two types, product of two types), or equal to some other type. If unification succeeds, types may be finalized. This sets all free types to unit and does one additional check (the "occurs check") ... .

Reply:
I like this narrative, but the problem is that we may have types with no free subtypes which are nontheless "incomplete". This happens if the subtypes get unified with something else so their free variables go away... but there is no way for the parent to be informed of this, so the parent continues to be "non-final" in the sense of having no cached TMR etc.

So I unfortunately need this awkward notion of "finalization" which is often basically a no-op, and which doesn't correspond simply to "has no free children".

Reply:
Let me revisit this. I think maybe I can hide these details from the user and present a simple "finalized means no free children" model.

Reply:
I like your trichotomy of incomplete types, complete types and finalized types. Types may be incomplete. Complete types have no free variables. Finalized types have cached information.